packages/react-meteor-data/useTracker.js

```diff
@@ -0,0 +1,155 @@
+  );
+}
+
+// inspired by https://github.com/facebook/react/blob/
+// 34ce57ae751e0952fd12ab532a3e5694445897ea/packages/
+// react-reconciler/src/ReactFiberHooks.js#L307-L354
+// used to replicate dep change behavior and stay consistent
+// with React.useEffect()
+function areHookInputsEqual(nextDeps, prevDeps) {
+  if (prevDeps === null || prevDeps === undefined || !Array.isArray(prevDeps)) {
+    return false;
+  }
+
+  if (!Array.isArray(nextDeps)) {
+    if (Meteor.isDevelopment) {
+      warn(
+        'Warning: useTracker expected an dependency value of '
+        + `type array but got type of ${typeof nextDeps} instead.`
+      );
+    }
+    return false;
+  }
+
+  const len = nextDeps.length;
+
+  if (prevDeps.length !== len) {
+    return false;
+  }
+
+  for (let i = 0; i < len; i++) {
+    if (!is(nextDeps[i], prevDeps[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Used to create a forceUpdate from useReducer. Forces update by
+// incrementing a number whenever the dispatch method is invoked.
+const fur = x => x + 1;
+
+function useTracker(reactiveFn, deps) {
+  const { current: refs } = useRef({});
+
+  const [, forceUpdate] = useReducer(fur, 0);
+
+  const dispose = () => {
+    if (refs.computation) {
+      refs.computation.stop();
+      refs.computation = null;
+    }
+  };
+
+  // this is called like at componentWillMount and componentWillUpdate equally
+  // in order to support render calls with synchronous data from the reactive computation
+  // if prevDeps or deps are not set areHookInputsEqual always returns false
+  // and the reactive functions is always called
+  if (!areHookInputsEqual(deps, refs.previousDeps)) {
+    // if we are re-creating the computation, we need to stop the old one.
+    dispose();
+
+    // store the deps for comparison on next render
+    refs.previousDeps = deps;
+
+    // Use Tracker.nonreactive in case we are inside a Tracker Computation.
+    // This can happen if someone calls `ReactDOM.render` inside a Computation.
+    // In that case, we want to opt out of the normal behavior of nested
+    // Computations, where if the outer one is invalidated or stopped,
+    // it stops the inner one.
+    refs.computation = Tracker.nonreactive(() => (
+      Tracker.autorun((c) => {
+        const runReactiveFn = () => {
+          const data = reactiveFn();
+          if (Meteor.isDevelopment) checkCursor(data);
+          refs.trackerData = data;
+        };
+
+        if (c.firstRun) {
+          // This will capture data synchronously on first run (and after deps change).
+          // Additional cycles will follow the normal computation behavior.
+          runReactiveFn();
+        } else {
+          // If deps are falsy, stop computation and let next render handle reactiveFn.
+          if (refs.previousDeps !== null && refs.previousDeps !== undefined
+            && !Array.isArray(refs.previousDeps)) {
+            dispose();
```

Comment:
I don't think I get that conditional : if previousDeps are neither null, undefined nor an array ? What would they be then ? Meaning, I don't get in which cases we go in that code branch currently ?
(and if the code is correct, then it's the comment that if off ?)

Reply:
Maybe it could be helpful to explain why we're looking at previousDeps rather than deps here ?

Reply:
Valid types for deps are `null`, `undefined` or an array. Previously, I thought we were checking for all "falsy" values, but I think the react code only checks for the listed types (as does the above deps compare check). The comment is out of date.

I'm looking at `previousDeps` because the ref is at this point always the update to date version of the deps. This closure is always created right after deps change in response to that change. We could probably just check the enclosed `deps` instead of `previousDeps` - it doesn't really matter because they should always be the same inside this closure.

Reply:
I fixed the logical error, and added some better comments

Reply:
I fixed the logical error, and also took a harder look at it. It must be an optimization. The null and undefined checks must be early exits (as @menelike suggested) to avoid calling Array.isArray (maybe because that's often polyfilled), since Array.isArray would actually be enough on it's own, logically speaking.